## Python Fundamentals - Exercise 4.4

## 01. Importing Libraries & Paths/Datasets

In [40]:
#Library Imports
import pandas as pd
import numpy as np
import os

In [41]:
#Path Shortcut (Copied shortcut even further in since we're using orders and products dataset)
path = r'/Users/puneet/Desktop/Instacart Basket Analysis 08-2025/02 Data/Original Data' 

In [42]:
#Orders dataset integration
df_ords = pd.read_csv(os.path.join(path, 'orders.csv'), index_col = False)

In [43]:
#Products dataset integration
df_prods = pd.read_csv(os.path.join(path, 'products.csv'), index_col = False)

## 02.Data Wrangling

- Dropping Columns - df.drop() function
    - Ex: df.drop(columns = ['variable'])

In [44]:
#Drop "eval_set" column from orders dataset
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

The above shows results of removal. ACtual removal of the column from dataframe requires additional code 

In [45]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [46]:
#Actual removal of column from dataframe
df_ords = df_ords.drop(columns = ['eval_set'])

"df_ords_2" also an option, but consider making new dataframe takes up space

## -

- Value counts function, to show missing values (Nan in Python)
    - df['variable'].value_counts(dropna = False)
    - dropna, means doesn't drop any missing values

In [47]:
#Counting missing values from 'days since prior order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

## -

- Renaming Columns - df.rename() function
    - Ex: df.rename(columns = {'old_name' : 'new_name'}, inplace = True)

In [48]:
#Change 'order_dow' to 'orders_day_of_the_week'
df_ords.rename(columns = {'order_dow' : 'orders_day_of_the_week'}, inplace = True)

In [49]:
#Check to see if column name changed, using head() function
df_ords.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## -

- Changing Variables Data Types - astype() function, change to 'str' = string
    - Ex: Want to change 'order_id' to string because we don't need descriptive statistics, only need it to describe orders and users
    - df['variable'] = df['variable'].astype('str')

In [50]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [51]:
#Check datatype of column now
df_ords['order_id'].dtype

dtype('O')

'0' means object in padas, pandas version of a string

'dtype()' is singular column vs. 'dtypes()' is plural columns. Distinction is important

## -

- Transposing Data (Switching rows and columns, vise versa)
    - First add departments.csv file. Then transpose it ('T' function)

In [52]:
#Add departments dataset 
df_dep = pd.read_csv(os.path.join(path,'departments.csv'),index_col = False)

In [53]:
#Check columns
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [54]:
#Transpose rows and columns using T function
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [55]:
#Actually changing dataframe to new transposed version (And renamed to df_dep_T)
df_dep_t = df_dep.T

In [56]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## -

Headers to colummns are wrong. Need to add an index. Ex: df.reset.index()

In [57]:
#Resetting index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

Next steps are:
1. Create new header
2. Remove first row in dataframe
3. Add new header

In [58]:
#Take the first row of the df_dep_t for the header
new_header = df_dep_t.iloc[0]

df.iloc[] function takes the index number of rows or columns. "Loc-ating and i-nteger". Python starts indexing from zero!

In [59]:
new_header

0    department
Name: department_id, dtype: object

In [60]:
#Create new dataframe without the first row
df_dep_t_new = df_dep_t[1:]

In [61]:
#New dataframe
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

First row was removed. Now correct dataframe is showing. Next step is adding teh header back

In [62]:
#New header is just a list of columns
df_dep_t_new.columns = new_header

In [63]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 03. Data Dictionary

- Create data dictionary using department_id column from departments.csv and assign to products.csv dataset
- First need to turn department dataframe into a data dictionary

In [64]:
#Turn dataframe into dictionary - to_dict() function
data_dict = df_dep_t_new.to_dict('index')

In [65]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [66]:
#Check products.csv top 5 entries
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Cross reference dictionary with the department_id of first row (Chocolate Sandwhich Cookie)

In [67]:
#Check first row with data dictionary
print(data_dict.get('19'))

{'department': 'snacks'}


We now know 'Chocolate Sandwhich Cookies' fall under the 'Snacks' department

## -

## 04. Subsetting

- Creating smaller dataset from a whole dataset (df_prods) based on a filter
- We're creating subset of df_prods that only contains data from snacks department

In [68]:
df_snacks = df_prods[df_prods['department_id']==19]

In [69]:
#Indexing element breakdown
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

This shows you what Python is looking at. Showing all values and assinging True or False based on if it's in the snacks department

In [70]:
#This next line wraps the initial dataframe to the logic we just wrote above 
#Now this dataframe only shows true values of department 19 values
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

We labled this earlier as df_snacks now

In [71]:
#First 5 rows of df_snacks, our new snacks only dataframe
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

Can also use the 'loc' function to get same result

In [72]:
#loc function example 1
df_snacks_2 = df_prods.loc[df_prods['department_id']==19]

In [73]:
#loc functione xample 2, you can check multiple values at the same time (.isin([17,18,19]))
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

## -

## 05. Exporting Dataframes

In [74]:
#Creating new path shortcut (original from up goes down to the 'Original' folder
path_2 = r'/Users/puneet/Desktop/Instacart Basket Analysis 08-2025'

In [75]:
#Export command for new .csv file
df_ords.to_csv(os.path.join(path_2, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

## -

## Exercise 4.4

2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [79]:
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [81]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0  2539329        1             1                       2                  8   
1  2398795        1             2                       3                  7   
2   473747        1             3                       3                 12   
3  2254736        1             4                       4                  7   
4   431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [82]:
#Changing 'user_id' column to string because numeric values don't matter
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [85]:
#Check user_id column for value change
df_ords['user_id'].dtype

dtype('O')

'user_id' is now a string 

3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [91]:
df_ords.head()

order_id user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0  2539329       1             1                       2                  8   
1  2398795       1             2                       3                  7   
2   473747       1             3                       3                 12   
3  2254736       1             4                       4                  7   
4   431534       1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [93]:
#Change name of 'order_number' to 'customer_order_count'
df_ords.rename(columns = {'order_number': 'customer_order_count'}, inplace = True)

In [94]:
df_ords.dtypes

order_id                   object
user_id                    object
customer_order_count        int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

Column name changed

4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [95]:
#Count values for 'order_hour_of_day' column
df_ords['order_hour_of_day'].value_counts (dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

10 AM is the busiest order hour of the day. The hours between 9AM-4PM are the busiest throughout the day

5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [102]:
#Data dictionary already established in exercise earlier, looking for value of 4
print(data_dict.get('4'))

{'department': 'produce'}


A department_id of '4' means it falls under 'Produce'

6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [103]:
#Locate breakfast items in data dictionary first
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [104]:
#Breakfast falls under department_id of 14

In [110]:
#Create Breakfast subset
df_breakfast = df_prods[df_prods['department_id']==14]

In [107]:
#Check subset
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [108]:
#Check data dictionary for dinner department_ids
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [109]:
#Need department_id 5,7,12, and 20

In [112]:
#Create dinner party subset
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [115]:
#Check dinner party subset
df_dinner_party.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [117]:
#Check rows and column count of dinner party dataframe subset above
df_dinner_party.shape

(7650, 5)

The dinner party dataframe has 7,650 rows

9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [126]:
#Create dataframe where user_id 1 is showing
df_user1 = df_ords[df_ords['user_id']== '1']

In [129]:
#Show user1 customer dataframe
df_user1

order_id user_id  customer_order_count  orders_day_of_the_week  \
0   2539329       1                     1                       2   
1   2398795       1                     2                       3   
2    473747       1                     3                       3   
3   2254736       1                     4                       4   
4    431534       1                     5                       4   
5   3367565       1                     6                       2   
6    550135       1                     7                       1   
7   3108588       1                     8                       1   
8   2295261       1                     9                       1   
9   2550362       1                    10                       4   
10  1187899       1                    11                       4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [131]:
df_user1.describe()

customer_order_count  orders_day_of_the_week  order_hour_of_day  \
count             11.000000               11.000000          11.000000   
mean               6.000000                2.636364          10.090909   
std                3.316625                1.286291           3.477198   
min                1.000000                1.000000           7.000000   
25%                3.500000                1.500000           7.500000   
50%                6.000000                3.000000           8.000000   
75%                8.500000                4.000000          13.000000   
max               11.000000                4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [132]:
#User1 has placed 11 orders. 
#They have only placed orders between either Sun-Wed, or Mon-Thursday. 
#They have only placed orders between the hours of 7AM and 4PM
#User1 places an order on average every 19 days, and hasn't gone over a month between orders

12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

Already done in command line 75
- df_ords.to_csv(os.path.join(path_2, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [133]:
df_dep_t_new.to_csv(os.path.join(path_2, '02 Data', 'Prepared Data','departments_wrangled.csv'))